packages/opentelemetry-context-async-hooks/test/AsyncHooksContextManager.test.ts

```diff
@@ -104,6 +104,166 @@ describe('AsyncHooksContextManager', () => {
+      let done = false;
+      await contextManager.withAsync(Context.ROOT_CONTEXT, async () => {
+        done = true;
+      });
+
+      assert.ok(done);
+    });
+
+    it('should rethrow errors', async () => {
+      contextManager.disable();
+      let done = false;
+      const err = new Error();
+
+      try {
+        await contextManager.withAsync(Context.ROOT_CONTEXT, async () => {
+          throw err;
+        });
+      } catch (e) {
+        assert.ok(e === err);
+        done = true;
+      }
+
+      assert.ok(done);
+    });
+
+    it('should finally restore an old scope', async () => {
+      const scope1 = '1' as any;
+      const scope2 = '2' as any;
+      let done = false;
+
+      await contextManager.withAsync(scope1, async () => {
+        assert.strictEqual(contextManager.active(), scope1);
+        await contextManager.withAsync(scope2, async () => {
+          assert.strictEqual(contextManager.active(), scope2);
+          done = true;
+        });
+        assert.strictEqual(contextManager.active(), scope1);
+      });
+
+      assert.ok(done);
+    });
+  });
+
+  describe('.withAsync/with()', () => {
+    it('with() inside withAsync() should correctly restore context', async () => {
+      const scope1 = '1' as any;
+      const scope2 = '2' as any;
+      let done = false;
+
+      await contextManager.withAsync(scope1, async () => {
+        assert.strictEqual(contextManager.active(), scope1);
+        contextManager.with(scope2, () => {
+          assert.strictEqual(contextManager.active(), scope2);
+          done = true;
+        });
+        assert.strictEqual(contextManager.active(), scope1);
+      });
+
+      assert.ok(done);
+    });
+
+    it('withAsync() inside with() should correctly restore conxtext', done => {
+      const scope1 = '1' as any;
+      const scope2 = '2' as any;
+
+      contextManager.with(scope1, async () => {
+        assert.strictEqual(contextManager.active(), scope1);
+        await contextManager.withAsync(scope2, async () => {
+          assert.strictEqual(contextManager.active(), scope2);
+        });
+        assert.strictEqual(contextManager.active(), scope1);
+        return done();
+      });
+      assert.strictEqual(contextManager.active(), Context.ROOT_CONTEXT);
+    });
+
+    it('not awaited withAsync() inside with() should not restore context', done => {
+      const scope1 = '1' as any;
+      const scope2 = '2' as any;
+
+      contextManager.with(scope1, () => {
+        assert.strictEqual(contextManager.active(), scope1);
+        contextManager
+          .withAsync(scope2, async () => {
+            assert.strictEqual(contextManager.active(), scope2);
+          })
+          .then(() => {
+            assert.strictEqual(contextManager.active(), scope1);
+            return done();
+          });
+        // in this case the current scope is 2 since we
+        // didnt waited the withAsync call
+        assert.strictEqual(contextManager.active(), scope2);
```

Comment:
It is definitely confusing why scope 2 is here and not scope 1, and I don't think this is the behavior most people would expect, but if there's nothing we can do about it it's probably ok.

Reply:
to be sure the async is working I would add there a simple `setTimeout` as additional test

Reply:
@obecny You are right, i've added it